### PR TITLE
Create Generic Workflows for Publishing Packages

### PR DESCRIPTION
**Pull request by takameyer**
## What, How & Why?
This is based on the "prepare-release.yml" and "publish-release.yml"
There currently the release process for packages is manual and I'm tired of doing it this way 😂 
NOTE: Currently not sure how to test this out without actually trying it out after being merged to main.

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 📝 Update `COMPATIBILITY.md`
* [ ] 🚦 Tests
* [ ] 🔀 Executed flexible sync tests locally if modifying flexible sync
* [ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated

### CODE REVIEW

File: .github/workflows/prepare-package-release.yml [hunk 1]
```diff
@@ -25,3 +25,68 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - run: echo "Preparing release for ${{inputs.package}}"
+      - uses: kanga333/variable-mapper@master
```
Comment (kraenhansen) [inline .github/workflows/prepare-package-release.yml:28]:
```suggestion
      - uses: kanga333/variable-mapper@v0.3.0
```

File: .github/workflows/prepare-package-release.yml [hunk 2]
```diff
@@ -25,3 +25,68 @@ jobs:
...
+        run: |
+          jq '.version="${{ steps.update-changelog.outputs.new-version }}"' package.json > package.tmp
+          mv package.tmp package.json
+
+      - name: Install Dependencies and update package-lock.json
+        run: |
+          npx lerna bootstrap --scope=${{ env.package-name }} --include-dependencies
+
+      - name: Create Release PR
+        uses: peter-evans/create-pull-request@7380612b49221684fefa025244f2ef4008ae50ad #! 3.10.1
```
Comment (kraenhansen) [inline .github/workflows/prepare-package-release.yml:86]: We might as well use the latest version?

```suggestion
        uses: peter-evans/create-pull-request@v4
```

File: .github/workflows/publish-package-release.yml [hunk 3]
```diff
@@ -29,4 +29,130 @@ jobs:
   publish-release:
     runs-on: ubuntu-latest
     steps:
-      - run: echo "Publishing release for ${{inputs.package}}"
+    - run: echo "Publishing release for ${{inputs.package}}"
+    - uses: kanga333/variable-mapper@master
```
Comment (kraenhansen) [inline .github/workflows/publish-package-release.yml:33]:
```suggestion
    - uses: kanga333/variable-mapper@v0.3.0
```

File: .github/workflows/publish-package-release.yml [hunk 4]
```diff
@@ -29,4 +29,130 @@ jobs:
...
+      uses: actions/checkout@v3
+      with:
+        submodules: recursive
+        ref: ${{ steps.find-pull-request.outputs.base-ref }}
+
+    - name: Extract Changelog
+      run: node scripts/extract-changelog.js ${{ steps.get-version.outputs.version }} ${{ github.workspace }}/${{env.package-path}}/CHANGELOG.md > EXTRACTED_CHANGELOG.md
+
+    - name: Create Github Release
+      uses: ncipollo/release-action@10c84d509b28aae3903113151bfd832314964f2e
```
Comment (kraenhansen) [inline .github/workflows/publish-package-release.yml:120]:
```suggestion
      uses: ncipollo/release-action@v1
```

File: .github/workflows/publish-package-release.yml [hunk 5]
```diff
@@ -29,4 +29,130 @@ jobs:
...
+      # This job might fail due to failed markdown-to-slack conversion.
+      continue-on-error: true
+
+    - name: Update Changelog
+      working-directory: ${{ github.workspace }}/${{env.package-path}}
+      run: ${{ github.workspace }}/scripts/changelog-header.sh
+
+    - name: Create vNext PR
+      id: vnext-pr
+      uses: peter-evans/create-pull-request@7380612b49221684fefa025244f2ef4008ae50ad
```
Comment (kraenhansen) [inline .github/workflows/publish-package-release.yml:144]:
```suggestion
      uses: peter-evans/create-pull-request@v4
```

File: .github/workflows/publish-package-release.yml [hunk 6]
```diff
@@ -29,4 +29,130 @@ jobs:
...
+      with:
+        branch: prepare-vnext
+        title: Prepare for ${{env.release-title}} vNext
+        body: Update Changelog for vNext
+        delete-branch: true
+        commit-message: Prepare for vNext
+        base: ${{ steps.find-pull-request.outputs.base-ref }}
+
+    - name: Merge Pull Request
+      uses: juliangruber/merge-pull-request-action@8a13f2645ad8b6ada32f829b2fae9c0955a5265d
```
Comment (kraenhansen) [inline .github/workflows/publish-package-release.yml:154]:
```suggestion
      uses: juliangruber/merge-pull-request-action@v1
```

File: scripts/build-ios.sh [hunk 7]
```diff
@@ -88,11 +88,11 @@ mkdir -p build
 pushd build
 
 # Configure CMake project
-SDKROOT="${SDK_ROOT_OVERRIDE:-/Applications/Xcode_13.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/}" cmake "$PROJECT_ROOT" -GXcode \
+SDKROOT="${SDK_ROOT_OVERRIDE:-/Applications/Xcode_14.2.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/}" cmake "$PROJECT_ROOT" -GXcode \
```
Comment (kraenhansen) [inline scripts/build-ios.sh:91]: Would it make sense to make a separate PR for this? In case we need to revert.

File: scripts/extract-changelog.js [hunk 8]
```diff
@@ -33,7 +33,8 @@ function extractRelease(changelog, expectedVersion) {
 }
 
 if (require.main === module) {
-  const changelogPath = path.resolve(__dirname, "../CHANGELOG.md");
+  const changelogPathArg = process.argv[3];
+  const changelogPath = path.resolve(__dirname, changelogPathArg ? changelogPathArg : "../CHANGELOG.md");
```
Comment (kraenhansen) [inline scripts/extract-changelog.js:37]:
```suggestion
  const changelogPath = path.resolve(__dirname, process.argv[3] || "../CHANGELOG.md");
```

**Review by kraenhansen — APPROVED** [review]
I have only minor suggestions. I'd like to see this tested, either while a PR or shortly thereafter, so we don't expect to rely on an untested workflow.

Should the alternative workflows to prepare and publish be deleted once this is merged or as a part of this PR?

**Comment by takameyer** [comment]
> Should the alternative workflows to prepare and publish be deleted once this is merged or as a part of this PR?

@kraenhansen To reduce the complexity of this workflow (mostly with finding the correct working directories), I excluded the release of `realm` from here.  I think later on when `realm` is part of `packages` we should manage the release of realm through these workflows as well.

File: .github/workflows/publish-package-release.yml [hunk 9]
```diff
@@ -29,4 +29,130 @@ jobs:
...
+    - uses: kanga333/variable-mapper@master
+      with:
+        key: "${{inputs.package}}"
+        map: |
+          {
+            "@realm/babel-plugin": {
+              "package-path": "packages/babel-plugin",
+              "package-name": "@realm/babel-plugin",
+              "version-prefix": "babel-plugin-",
+              "release-title": "Realm Babel Plugin"
```
Comment (kraenhansen) [inline .github/workflows/publish-package-release.yml:42]: Admittedly, not super important, but I guess there's a precedence for naming these CAPITALIZED_SNAKE_CASE.

File: scripts/build-ios.sh (hunk 7 above)
Comment (takameyer) [reply scripts/build-ios.sh:91]: This needs to be reverted.  Result of too much multi-tasking

**Review by kneth — APPROVED** [review]

**Review by kraenhansen — APPROVED** [review]
